File: HMCLCore/src/main/java/org/jackhuang/hmcl/util/Logging.java
/*
 * Hello Minecraft! Launcher
 * Copyright (C) 2021  huangyuhui <huanghongxun2008@126.com> and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.jackhuang.hmcl.util;

import org.jackhuang.hmcl.util.io.IOUtils;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.logging.*;

/**
 * @author huangyuhui
 */
public final class Logging {
    private Logging() {
    }

    public static final Logger LOG = Logger.getLogger("HMCL");
    private static final ByteArrayOutputStream storedLogs = new ByteArrayOutputStream(IOUtils.DEFAULT_BUFFER_SIZE);

    private static volatile String[] accessTokens = new String[0];

    public static synchronized void registerAccessToken(String token) {
        final String[] oldAccessTokens = accessTokens;
        final String[] newAccessTokens = Arrays.copyOf(oldAccessTokens, oldAccessTokens.length + 1);

        newAccessTokens[oldAccessTokens.length] = token;

        accessTokens = newAccessTokens;
    }

    public static String filterForbiddenToken(String message) {
        for (String token : accessTokens)
            message = message.replace(token, "<access token>");
        return message;
    }

    public static void start(Path logFolder) {
        LOG.setLevel(Level.ALL);
        LOG.setUseParentHandlers(false);
        LOG.setFilter(record -> {
            record.setMessage(format(record));
            return true;
        });

        DefaultFormatter formatter = new DefaultFormatter();
        try {
            if (Files.isRegularFile(logFolder))
                Files.delete(logFolder);

            Files.createDirectories(logFolder);
            FileHandler fileHandler = new FileHandler(logFolder.resolve("hmcl.log").toAbsolutePath().toString());
            fileHandler.setLevel(Level.FINEST);
            fileHandler.setFormatter(formatter);
            fileHandler.setEncoding("UTF-8");
            LOG.addHandler(fileHandler);
        } catch (IOException e) {
            System.err.println("Unable to create hmcl.log\n" + StringUtils.getStackTrace(e));
        }

        ConsoleHandler consoleHandler = new ConsoleHandler();
        consoleHandler.setFormatter(formatter);
        consoleHandler.setLevel(Level.FINER);
        LOG.addHandler(consoleHandler);

        StreamHandler streamHandler = new StreamHandler(storedLogs, formatter) {
            @Override
            public synchronized void publish(LogRecord record) {
                super.publish(record);
                flush();
            }
        };
        try {
            streamHandler.setEncoding("UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        streamHandler.setLevel(Level.ALL);
        LOG.addHandler(streamHandler);
    }

    public static void initForTest() {
        LOG.setLevel(Level.ALL);
        LOG.setUseParentHandlers(false);

        ConsoleHandler consoleHandler = new ConsoleHandler();
        consoleHandler.setFormatter(new DefaultFormatter());
        consoleHandler.setLevel(Level.FINER);
        LOG.addHandler(consoleHandler);
    }

    public static byte[] getRawLogs() {
        return storedLogs.toByteArray();
    }

    public static String getLogs() {
        try {
            return storedLogs.toString("UTF-8");
        } catch (UnsupportedEncodingException e) {
            throw new InternalError(e);
        }
    }

    private static final DateTimeFormatter TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm:ss").withZone(ZoneId.systemDefault());

    private static String format(LogRecord record) {
        String message = filterForbiddenToken(record.getMessage());

        StringBuilder builder = new StringBuilder(128 + message.length());
        builder.append('[');
        TIME_FORMATTER.formatTo(Instant.ofEpochMilli(record.getMillis()), builder);
        builder.append(']');

        builder.append(" [")
                .append(record.getSourceClassName())
                .append('.')
                .append(record.getSourceMethodName())
                .append('/')
                .append(record.getLevel().getName())
                .append("] ")
                .append(message)
                .append('\n');


        Throwable thrown = record.getThrown();
        if (thrown == null) {
            return builder.toString();
        } else {
            StringWriter writer = new StringWriter(builder.length() + 2048);
            writer.getBuffer().append(builder);
            try (PrintWriter printWriter = new PrintWriter(writer)) {
                thrown.printStackTrace(printWriter);
            }

            return writer.toString();
        }
    }

    private static final class DefaultFormatter extends Formatter {
        @Override
        public String format(LogRecord record) {
            return record.getMessage();
        }
    }
}

File: HMCLCore/src/main/kotlin/org/jackhuang/hmcl/util/Logging.kt
/*
 * Hello Minecraft! Launcher.
 * Copyright (C) 2017  huangyuhui <huanghongxun2008@126.com>
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see {http://www.gnu.org/licenses/}.
 */
@file:JvmName("HMCLog")
package org.jackhuang.hmcl.util

import java.io.ByteArrayOutputStream
import java.io.PrintWriter
import java.text.SimpleDateFormat
import java.util.*
import java.util.logging.*
import java.util.logging.Formatter
import javafx.scene.paint.Color
import java.util.regex.Pattern

val LOG = Logger.getLogger("HMCL").apply {
    level = Level.FINER
    useParentHandlers = false
    addHandler(FileHandler("hmcl.log").apply {
        level = Level.FINER
        formatter = DefaultFormatter
    })
    addHandler(ConsoleHandler().apply {
        level = Level.FINER
        formatter = DefaultFormatter
    })
}

val DEFAULT_DATE_FORMAT = SimpleDateFormat("HH:mm:ss")

internal object DefaultFormatter : Formatter() {
    override fun format(record: LogRecord): String {
        var s: String = "[${DEFAULT_DATE_FORMAT.format(Date(record.millis))}] [${record.sourceClassName}.${record.sourceMethodName}/${record.level.name}] ${record.message}\n"
        val builder = ByteArrayOutputStream()
        if (record.thrown != null)
            PrintWriter(builder).use(record.thrown::printStackTrace)
        s += builder.toString()
        return s
    }

}

/**
 *
 * @author huangyuhui
 */
enum class Log4jLevel constructor(val level: Int, val color: Color) {

    FATAL(1, Color.RED),
    ERROR(2, Color.RED),
    WARN(3, Color.ORANGE),
    INFO(4, Color.BLACK),
    DEBUG(5, Color.BLUE),
    TRACE(6, Color.BLUE),
    ALL(2147483647, Color.BLACK);

    fun lessOrEqual(level: Log4jLevel): Boolean {
        return this.level <= level.level
    }

    companion object {

        val MINECRAFT_LOGGER = Pattern.compile("\\[(?<timestamp>[0-9:]+)\\] \\[[^/]+/(?<level>[^\\]]+)\\]")
        val MINECRAFT_LOGGER_CATEGORY = Pattern.compile("\\[(?<timestamp>[0-9:]+)\\] \\[[^/]+/(?<level>[^\\]]+)\\] \\[(?<category>[^\\]]+)\\]")
        val JAVA_SYMBOL = "([a-zA-Z_$][a-zA-Z\\d_$]*\\.)+[a-zA-Z_$][a-zA-Z\\d_$]*"

        fun guessLevel(line: String): Log4jLevel? {
            var level: Log4jLevel? = null
            val m = MINECRAFT_LOGGER.matcher(line)
            if (m.find()) {
                // New style logs from log4j
                val levelStr = m.group("level")
                if (null != levelStr)
                    when (levelStr) {
                        "INFO" -> level = INFO
                        "WARN" -> level = WARN
                        "ERROR" -> level = ERROR
                        "FATAL" -> level = FATAL
                        "TRACE" -> level = TRACE
                        "DEBUG" -> level = DEBUG
                        else -> {
                        }
                    }
                val m2 = MINECRAFT_LOGGER_CATEGORY.matcher(line)
                if (m2.find()) {
                    val level2Str = m2.group("category")
                    if (null != level2Str)
                        when (level2Str) {
                            "STDOUT" -> level = INFO
                            "STDERR" -> level = ERROR
                        }
                }
            } else {
                if (line.contains("[INFO]") || line.contains("[CONFIG]") || line.contains("[FINE]")
                        || line.contains("[FINER]") || line.contains("[FINEST]"))
                    level = INFO
                if (line.contains("[SEVERE]") || line.contains("[STDERR]"))
                    level = ERROR
                if (line.contains("[WARNING]"))
                    level = WARN
                if (line.contains("[DEBUG]"))
                    level = DEBUG
            }
            return if (line.contains("overwriting existing")) FATAL else level

            /*if (line.contains("Exception in thread")
                || line.matches("\\s+at " + JAVA_SYMBOL)
                || line.matches("Caused by: " + JAVA_SYMBOL)
                || line.matches("([a-zA-Z_$][a-zA-Z\\d_$]*\\.)+[a-zA-Z_$]?[a-zA-Z\\d_$]*(Exception|Error|Throwable)")
                || line.matches("... \\d+ more$"))
            return ERROR;*/
        }

        fun isError(a: Log4jLevel?): Boolean {
            return a?.lessOrEqual(ERROR) ?: false
        }

        fun mergeLevel(a: Log4jLevel?, b: Log4jLevel?): Log4jLevel? {
            return if (a == null) b
                    else if (b == null) a
                    else if (a.level < b.level) a else b
        }
    }

}

fun guessLogLineError(log: String) = Log4jLevel.isError(Log4jLevel.guessLevel(log))